packages/wrangler/src/module-collection.ts

```diff
@@ -1,23 +1,87 @@
-import type { CfModule, CfScriptFormat } from "./api/worker";
+import globToRegExp from "glob-to-regexp";
+import type { CfModule, CfModuleType, CfScriptFormat } from "./api/worker";
+import type { Config, ConfigModuleRuleType } from "./config";
 import type esbuild from "esbuild";
 
+const RuleTypeToModuleType: Record<ConfigModuleRuleType, CfModuleType> = {
+  ESModule: "esm",
+  CommonJS: "commonjs",
+  CompiledWasm: "compiled-wasm",
+  Data: "buffer",
+  Text: "text",
+};
+
 // This is a combination of an esbuild plugin and a mutable array
 // that we use to collect module references from source code.
 // There will be modules that _shouldn't_ be inlined directly into
 // the bundle. (eg. wasm modules, some text files, etc). We can include
-// those files as modules in the multi part forker form upload. This
+// those files as modules in the multi part worker form upload. This
 // plugin+array is used to collect references to these modules, reference
 // them correctly in the bundle, and add them to the form upload.
 
+export const DEFAULT_MODULE_RULES: Config["rules"] = [
+  { type: "Text", globs: ["**/*.txt", "**/*.html"] },
+  { type: "Data", globs: ["**/*.bin"] },
+  { type: "CompiledWasm", globs: ["**/*.wasm"] },
+];
+
 export default function makeModuleCollector(props: {
   format: CfScriptFormat;
+  rules?: Config["rules"];
 }): {
   modules: CfModule[];
   plugin: esbuild.Plugin;
 } {
+  const rules: Config["rules"] = [
+    ...(props.rules || []),
+    // eslint-disable-next-line @typescript-eslint/no-non-null-assertion
+    ...DEFAULT_MODULE_RULES!,
+  ];
+
+  const completedRuleLocations: Record<string, number> = {};
+  let index = 0;
+  const rulesToRemove: Config["rules"] = [];
+  for (const rule of rules) {
+    if (rule.type in completedRuleLocations) {
+      if (rules[completedRuleLocations[rule.type]].fallthrough !== false) {
+        if (index < (props.rules || []).length) {
+          console.warn(
+            `The module rule at position ${index} (${JSON.stringify(
+              rule
+            )}) has the same type as a previous rule (at position ${
+              completedRuleLocations[rule.type]
+            }, ${JSON.stringify(
+              rules[completedRuleLocations[rule.type]]
+            )}). This rule will be ignored. To the previous rule, add \`fallthrough = true\` to allow this one to also be used, or \`fallthrough = false\` to silence this warning.`
+          );
+        } else {
+          console.warn(
+            `The default module rule ${JSON.stringify(
+              rule
+            )} has the same type as a previous rule (at position ${
+              completedRuleLocations[rule.type]
+            }, ${JSON.stringify(
+              rules[completedRuleLocations[rule.type]]
+            )}). This rule will be ignored. To the previous rule, add \`fallthrough = true\` to allow the default one to also be used, or \`fallthrough = false\` to silence this warning.`
+          );
+        }
```

Comment:
NIT: using an `isDefault` constant can half the code here.

```suggestion
        const isDefault = (index < (props.rules || []).length;
        console.warn(
          `The ${isDefault ? "default " : ""}module rule at position ${index} (${JSON.stringify(
              rule
          )}) has the same type as a previous rule (at position ${
            completedRuleLocations[rule.type]
          }, ${JSON.stringify(
            rules[completedRuleLocations[rule.type]]
          )}). This rule will be ignored. To the previous rule, add \`fallthrough = true\` to allow ${isDefault ? "the default" : "this"} one to also be used, or \`fallthrough = false\` to silence this warning.`
          );
```

Reply:
I honestly prefer the explicit branch for the clearer message here, especially because we don't have to save bytes. I'll parry your nit!